src/ssht_numba/numba_wrappers.py

```diff
@@ -287,3 +280,193 @@ def ssht_numba_series_eval(f_lm, s, L, delta, theta, phi):
+
+        for el in range(1, L):
+            ssht_dl_beta_risbo_half_table(
+                ffi.from_buffer(dl_dummy),
+                beta,
+                L,
+                dl_size,
+                el,
+                ffi.from_buffer(sqrt_tbl),
+                ffi.from_buffer(signs),
+            )
+            for i in range(-el, el + 1):
+                for j in range(-el, el + 1):
+                    dl_array[el, offset_m + i, offset_m + j] = dl_dummy[
+                        offset_m + i, offset_m + j
+                    ]
+
+    return dl_array
+
+
+class SSHTInputError(TypeError):
+    """Raise this when inputs are the wrong type"""
+
+    pass
+
+
+class SSHTSpinError(ValueError):
+    """Raise this when the spin is none zero but Reality is true"""
+
+    pass
+
+
+def sample_shape(L, method="MW"):
+    shapes = {
+        "MW": (L, 2 * L - 1),
+        "MW_pole": (L - 1, 2 * L - 1),
+        "MWSS": (L - 1, 2 * L - 1),
+        "GL": (L, 2 * L - 1),
+        "DH": (2 * L, 2 * L - 1),
+    }
+    try:
+        return shapes[method]
+    except KeyError:
+        raise SSHTInputError(
+            "Method is not recognised, Methods are: MW, MW_pole, MWSS, DH and GL"
+        )
+
+
+def ssht(f, L: int, spin=0, method="MW", real=False, forward=True, adjoint=False):
+    dl_method = 0
+
+    # Checks
+    if f.ndim != 2:
+        raise SSHTInputError("f must be 2D numpy array")
+
+    _methods = ["MW", "MW_pole", "MWSS", "DH", "GL"]
+    if not adjoint and method not in _methods:
+        raise SSHTInputError("Method is not recognised, Methods are: %s" % _methods)
+    elif adjoint and method not in ["MW", "MWSS"]:
+        raise SSHTInputError("Method is not recognised, Methods are: MW, MWSS")
+
+    if spin and real:
+        raise SSHTSpinError(
+            "Reality set to True and Spin is not 0. However, spin signals must be complex."
+        )
+
+    if f.dtype == np.float_ and not real:
+        warnings.warn(
+            "Real signal given but Reality flag is False. "
+            "Set Reality = True to improve performance"
+        )
+        f_new = f + 1j * np.zeros(sample_shape(L, method=method), dtype=np.float_)
+        f = f_new
+
+    if f.dtype == np.complex_ and real:
+        warnings.warn(
+            "Complex signal given but Reality flag is True. Ignoring complex component"
+        )
+        f_new = np.real(f)
+        f = f_new.copy(order="c")
+
+    modifier = "_real" if real else ""
+    if method.startswith("MW"):
+        _method = "mw"
+        if method == "MWSS":
+            modifier = "_ss" + modifier
+        elif method == "MW_pole":
+            modifier = modifier + "_pole"
+    else:
+        _method = method.lower()
+
+    fnc = getattr(
+        _ssht_cffi.lib,
+        f"ssht_{'core' if not adjoint else 'adjoint'}_{_method}_"
+        f"{'forward' if forward else 'inverse'}_sov{'_conv' if forward else ''}"
+        f"_sym{modifier}",
+    )
+
+    _out_shapes = {
```

Comment:
need to add the signatures for all these functions to ssht.h

Reply:
Can you give some more information? What would the signature for `forward_mw_complex` look like, for example?

Reply:
Ah yes, I can do that. On another note, @zacharymartinot, I'm wondering whether my approach will really work, because I am dynamically accessing the `lib.<funcname>` within the function here. It will "work", but not necessarily in `numba`, right?

Reply:
OK, we can move this discussion to another PR, since I have removed this whole function.